Added a solidity scripting tutorial (#370)
foundry-rs/book: src/tutorials/solidity-scripting.md

## Solidity Scripting

### Introduction

> Before solidity scripting was introduced we only had one way to deploy contracts in Foundry to either mainnet or any other chain which was [`forge create`](../reference/forge/forge-create.md), ideally this was not UX friendly and a bit limiting so the Foundry team came up with the next-gen concept known as solidity scripting.

Reviewer: Doesn't need to be exactly like this, but I found the sentence overly verbose

```suggestion
Solidity scripting is a way to declaratively deploy contracts using Solidity, instead of using the more limiting and less user friendly [`forge create`](../reference/forge/forge-create.md).
```

> Solidity scripting is just like the regular scripts you write when working with tools like Hardhat, what makes solidity scripting different is the fact that instead of writing scripts in for instance JavaScript like you would in Hardhat, you write them in solidity instead. Like everything else in Foundry, it's crazy fast, efficient and more reliable.

Reviewer: Again; doesn't need to be exactly like this, but I'd say the major differences are:

- That it is written in Solidity
- That you have access to the Foundry EVM backend, so you can do things like dry-running before actually deploying, and you have access to a fuller range of features (i.e. all cheatcodes)

```suggestion
Solidity scripts are like the scripts you write when working with tools like Hardhat; what makes Solidity scripting different is that they are written in Solidity instead of JavaScript, and they are run on the fast Foundry EVM backend, which provides dry-run capabilities.
```

### Set Up

Let’s try to deploy the NFT contract made in the solmate tutorial with solidity scripting. First of all, we would need to create a new Foundry project via:

```bash
forge init solidity-scripting
```

Since the NFT contract from the solmate tutorial inherits both `solmate` and `OpenZepplin` contracts, we’ll have to install them as dependencies by running:

```bash
# Enter the project
cd solidity-scripting

# Install Solmate and OpenZepplin contracts as dependencies
forge install Rari-Capital/solmate Openzeppelin/openzeppelin-contracts
```

Next, we have to delete the `Contract.sol` file in the `src` folder and create another file called `NFT.sol`. You can do this by running:

```bash
rm src/Contract.sol && touch src/NFT.sol && ls src
```

> ![Untitled](../images/solidity-scripting%20/set-up-commands.png)

Reviewer: Please provide a better alt text than untitled :smile:

    ) ERC721(_name, _symbol) {
        baseURI = _baseURI;
    }

    function mintTo(address recipient) public payable returns (uint256) {
        if (msg.value != MINT_PRICE) {
            revert MintPriceNotPaid();
        }
        uint256 newTokenId = ++currentTokenId;
        if (newTokenId > TOTAL_SUPPLY) {
            revert MaxSupply();
        }
        _safeMint(recipient, newTokenId);
        return newTokenId;
    }

    function tokenURI(uint256 tokenId)
        public
        view
        virtual
        override
        returns (string memory)
    {
        if (ownerOf(tokenId) == address(0)) {
            revert NonExistentTokenURI();
        }
        return
            bytes(baseURI).length > 0
                ? string(abi.encodePacked(baseURI, tokenId.toString()))
                : "";
    }

    function withdrawPayments(address payable payee) external onlyOwner {
        uint256 balance = address(this).balance;
        (bool transferTx, ) = payee.call{value: balance}("");
        if (!transferTx) {
            revert WithdrawTransfer();
        }
    }
}
```

Now, let’s try compiling our contract to make sure everything is in order.

```bash
forge build
```

> ![Untitled](../images/solidity-scripting%20/compile-successful.png)

Reviewer: Same as above

> If you get this, it's the contract compiled successfully.

Reviewer: Maybe hoist it above the image?

```suggestion
If your output looks like this, the contracts successfully compiled.
```

### Deploying our contract

> We’re going to deploy the `NFT` contract to the rinkeby testnet but to do this we’ll need some variables like a Rinkeby Rpc Url, the private key of an account that’s funded with Rinkeby Eth and an Etherscan key for the verification of the NFT contract.

Reviewer: ```suggestion
We’re going to deploy the `NFT` contract to the Rinkeby testnet, but to do this we’ll need to configure Foundry a bit, by setting things like a Rinkeby RPC URL, the private key of an account that’s funded with Rinkeby Eth, and an Etherscan key for the verification of the NFT contract.
```

> 💡 Note: You can get some Rinkeby testnet ETH [here](https://faucet.paradigm.xyz/) .

Once you have all that create a `.env` file and add the variables.

> You can create the `.env` file by running:
> 
> ```bash
> touch .env
> ```

Reviewer: ```suggestion

```

Doesn't matter much IMO - people can just create the file in whatever editor they want

Author: Sorry, I'm a sucker for details. haha

The .env file should follow this format:

```bash
RINKEBY_RPC_URL=
PRIVATE_KEY=
ETHERSCAN_KEY=
```

> Next, we have to create a folder and name it `script` then create a file in it called `NFT.s.sol`, the contents of NFT.s.sol should look like this:

Reviewer: ```suggestion
Next, we have to create a folder and name it `script` and create a file in it called `NFT.s.sol`. This is where we will create the deployment script itself.

The contents of `NFT.s.sol` should look like this:
```

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

import "forge-std/Script.sol";
import "../src/NFT.sol";

contract MyScript is Script {
    function run() external {
        vm.startBroadcast();

        NFT nft = new NFT("NFT_tutorial", "TUT", "baseUri");

        vm.stopBroadcast();
    }
}
```

> This is what a script in Foundry looks like, now let’s take a walkthrough of the code and what it actually means and does.

Reviewer: ```suggestion
Now let’s read through the code and figure out what it actually means and does.
```

```solidity
// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;
```

> Remember even if it’s a script it still works like a smart contract but is never deployed, so just like any other smart contract written in solidity the `pragma version` has to be specified.

Reviewer: ```suggestion
Remember even if it’s a script it still works like a smart contract, but is never deployed, so just like any other smart contract written in Solidity the `pragma version` has to be specified.
```

```solidity
import "forge-std/Script.sol";
import "../src/NFT.sol";
```

> Just like how in Foundry tests we import `forge-std/Test.sol` to get utilities we’ll in our tests, that’s how we import `forge-std/Script.sol` in our scripts to get utilities we’ll need while writing scripts.

Reviewer: ```suggestion
Just like we may import Forge Std to get testing utilities when writing tests, Forge Std also provides some scripting utilities that we import here.
```

The next line just imports the `NFT` contract.

```solidity
contract MyScript is Script {
```

> We create a contract called `MyScript` and it inherits `Script` which is from the file we imported earlier.

Reviewer: ```suggestion
We create a contract called `MyScript` and it inherits `Script` from Forge Std.
```

```solidity
 function run() external {
```

> For script smart contracts we only need one function and that’s the run function, think about it like the main function in a Hardhat script written in JavaScript. We deploy our contracts in this function.

Reviewer: ```suggestion
By default, scripts are executed by calling the function named `run`, our entrypoint.
```

```solidity
vm.startBroadcast();
```

> This Foundry cheat code is uses the address that calls the `Script` contract to create a transaction that can later be signed and sent on chain. In other words we’ll use it create the deployment transaction for the NFT contract.

Reviewer: ```suggestion
This is a special cheatcode that records calls and contract creations made by our main script contract. Later, we will broadcast these transactions to deploy our NFT contract.
```

```solidity
 NFT nft = new NFT("NFT_tutorial", "TUT", "baseUri");
```

> You should be familiar with this part, all it does is create an instance of our NFT contract which will be deployed once we run the script.

Reviewer: ```suggestion
Here we just create our NFT contract. Because we called `vm.startBroadcast()` before this line, the contract creation will be recorded by Forge, and as mentioned previously, we can broadcast the transaction to deploy the contract on-chain.
```

Now that you’re up to speed about what the script smart contract does, let’s run it.

You should have added the variables we mentioned earlier to the `.env` for this next part to work.

At the root of the project run:

```bash
# To load the variables in the .env file
source .env

# To deploy and verify our contract
forge script script/NFT.s.sol:MyScript --rpc-url $RINKEBY_RPC_URL  --private-key $PRIVATE_KEY --broadcast --verify --etherscan-api-key $ETHERSCAN_KEY -vvvv
```

> It’s going to take some time not up to a minute, at the bottom of your terminal you should see something like this:

Reviewer: ```suggestion
Forge is going to run our script and broadcast the transactions for us - this can take a little while, since Forge will also wait for the transaction receipts. You should see something like this after a minute or so:
```

> ![Untitled](../images/solidity-scripting%20/contract-verified.png)

Reviewer: Better alt text please :smile:

Author: I'm sorry about that. I work on it.
